File: client-api/src/test/java/org/ostelco/topup/api/resources/ProductsResourceTest.java
package org.ostelco.topup.api.resources;

import io.dropwizard.auth.AuthDynamicFeature;
import io.dropwizard.auth.AuthValueFactoryProvider;
import io.dropwizard.auth.oauth.OAuthCredentialAuthFilter;
import io.dropwizard.testing.junit.ResourceTestRule;
import io.vavr.control.Either;
import io.vavr.control.Option;
import org.glassfish.jersey.test.grizzly.GrizzlyWebTestContainerFactory;
import org.junit.Before;
import org.junit.ClassRule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.ostelco.prime.model.Price;
import org.ostelco.prime.model.Product;
import org.ostelco.topup.api.auth.AccessTokenPrincipal;
import org.ostelco.topup.api.auth.AccessTokenPrincipal;
import org.ostelco.topup.api.auth.OAuthAuthenticator;
import org.ostelco.topup.api.db.SubscriberDAO;
import org.ostelco.topup.api.util.AccessToken;

import java.util.Base64;
import java.util.List;
import java.util.Optional;
import javax.ws.rs.client.Entity;
import javax.ws.rs.core.GenericType;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import static java.util.Collections.emptyMap;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Products API tests.
 *
 */
public class ProductsResourceTest {

    private static final SubscriberDAO DAO = mock(SubscriberDAO.class);
    private static final OAuthAuthenticator AUTHENTICATOR = mock(OAuthAuthenticator.class);

    private final String email = "mw@internet.org";

    private final List<Product> products = io.vavr.collection.List.of(
            new Product("1", new Price(10, "NOK"), emptyMap(), emptyMap()),
            new Product("2", new Price(5, "NOK"), emptyMap(), emptyMap()),
            new Product("3", new Price(20, "NOK"), emptyMap(), emptyMap()))
            .toJavaList();
    private final String userInfo = Base64.getEncoder()
            .encodeToString((new String("{\n" +
                    "     \"issuer\": \"someone\",\n" +
                    "     \"email\": \"mw@internet.org\"\n" +
                    "}\n"))
                    .getBytes());

    @ClassRule
    public static final ResourceTestRule RULE = ResourceTestRule.builder()
            .addResource(new AuthDynamicFeature(
                    new OAuthCredentialAuthFilter.Builder<AccessTokenPrincipal>()
                            .setAuthenticator(AUTHENTICATOR)
                            .setPrefix("Bearer")
                            .buildAuthFilter()))
            .addResource(new AuthValueFactoryProvider.Binder<>(AccessTokenPrincipal.class))
            .addResource(new ProductsResource(DAO))
            .setTestContainerFactory(new GrizzlyWebTestContainerFactory())
            .build();

    @Before
    public void setUp()  throws Exception {
        when(AUTHENTICATOR.authenticate(anyString()))
            .thenReturn(Optional.of(new AccessTokenPrincipal(email)));
    }

    @Test
    public void getProducts() throws Exception {
        ArgumentCaptor<String> arg = ArgumentCaptor.forClass(String.class);

        when(DAO.getProducts(arg.capture())).thenReturn(Either.right(products));

        Response resp = RULE.target("/products")
                .request()
                .header("Authorization", String.format("Bearer %s", AccessToken.withEmail(email)))
                .header("X-Endpoint-API-UserInfo", userInfo)
                .get(Response.class);

        assertThat(resp.getStatus()).isEqualTo(Response.Status.OK.getStatusCode());
        assertThat(resp.getMediaType().toString()).isEqualTo(MediaType.APPLICATION_JSON);

        // assertThat and assertEquals is not working
        assertTrue(products.equals(resp.readEntity(new GenericType<List<Product>>() {})));
        assertThat(arg.getValue()).isEqualTo(email);
    }

    @Test
    public void purchaseProduct() throws Exception {
        ArgumentCaptor<String> arg1 = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<String> arg2 = ArgumentCaptor.forClass(String.class);

        final String sku = products.get(0).getSku();

        when(DAO.purchaseProduct(arg1.capture(), arg2.capture())).thenReturn(Option.none());

        Response resp = RULE.target(String.format("/products/%s", sku))
                .request()
                .header("Authorization", String.format("Bearer %s", AccessToken.withEmail(email)))
                .header("X-Endpoint-API-UserInfo", userInfo)
                .post(Entity.text(""));

        assertThat(resp.getStatus()).isEqualTo(Response.Status.CREATED.getStatusCode());
        assertThat(arg1.getValue()).isEqualTo(email);
        assertThat(arg2.getValue()).isEqualTo(sku);
    }
}

File: client-api/src/test/kotlin/org/ostelco/prime/client/api/resources/ProductsResourceTest.kt
package org.ostelco.prime.client.api.resources

import arrow.core.Either
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.nhaarman.mockito_kotlin.argumentCaptor
import io.dropwizard.auth.AuthDynamicFeature
import io.dropwizard.auth.AuthValueFactoryProvider
import io.dropwizard.auth.oauth.OAuthCredentialAuthFilter
import io.dropwizard.jackson.Jackson
import io.dropwizard.testing.junit.ResourceTestRule
import org.assertj.core.api.Assertions.assertThat
import org.glassfish.jersey.test.grizzly.GrizzlyWebTestContainerFactory
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.ClassRule
import org.junit.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.ostelco.prime.client.api.auth.AccessTokenPrincipal
import org.ostelco.prime.client.api.auth.OAuthAuthenticator
import org.ostelco.prime.client.api.store.SubscriberDAO
import org.ostelco.prime.client.api.util.AccessToken
import org.ostelco.prime.core.ApiError
import org.ostelco.prime.model.Price
import org.ostelco.prime.model.Product
import org.ostelco.prime.paymentprocessor.PaymentProcessor
import org.ostelco.prime.paymentprocessor.core.ProductInfo
import java.util.*
import java.util.Collections.emptyMap
import javax.ws.rs.client.Entity
import javax.ws.rs.core.GenericType
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

private val PAYMENT: PaymentProcessor = mock(PaymentProcessor::class.java)
class MockPaymentProcessor : PaymentProcessor by PAYMENT

/**
 * Products API tests.
 *
 */
class ProductsResourceTest {

    private val email = "mw@internet.org"

    private val products = listOf(
            Product("1", Price(10, "NOK"), emptyMap(), emptyMap()),
            Product("2", Price(5, "NOK"), emptyMap(), emptyMap()),
            Product("3", Price(20, "NOK"), emptyMap(), emptyMap()))

    private val userInfo = Base64.getEncoder().encodeToString(
            """|{
               |  "issuer": "someone",
               |  "email": "mw@internet.org"
               |}""".trimMargin()
                    .toByteArray())

    @Before
    @Throws(Exception::class)
    fun setUp() {
        `when`(AUTHENTICATOR.authenticate(ArgumentMatchers.anyString()))
                .thenReturn(Optional.of(AccessTokenPrincipal(email)))
    }

    @Test
    @Throws(Exception::class)
    fun getProducts() {
        val arg = argumentCaptor<String>()

        `when`<Either<ApiError, Collection<Product>>>(DAO.getProducts(arg.capture())).thenReturn(Either.right(products))

        val resp = RULE.target("/products")
                .request()
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .header("X-Endpoint-API-UserInfo", userInfo)
                .get(Response::class.java)

        assertThat(resp.status).isEqualTo(Response.Status.OK.statusCode)
        assertThat(resp.mediaType.toString()).isEqualTo(MediaType.APPLICATION_JSON)

        // assertThat and assertEquals is not working
        assertTrue(products == resp.readEntity(object : GenericType<List<Product>>() {

        }))
        assertThat(arg.firstValue).isEqualTo(email)
    }

    @Test
    @Throws(Exception::class)
    fun purchaseProduct() {
        val emailArg = argumentCaptor<String>()
        val skuArg = argumentCaptor<String>()
        val sourceIdArg = argumentCaptor<String>()
        val saveSourceArg = argumentCaptor<Boolean>()

        val sku = products[0].sku
        val sourceId = "amex"

        `when`(DAO.purchaseProduct(
                emailArg.capture(),
                skuArg.capture(),
                sourceIdArg.capture(),
                saveSourceArg.capture())).thenReturn(Either.right(ProductInfo(sku)))

        val resp = RULE.target("/products/$sku/purchase")
                .queryParam("sourceId", sourceId)
                .request()
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .header("X-Endpoint-API-UserInfo", userInfo)
                .post(Entity.text(""))

        assertThat(resp.status).isEqualTo(Response.Status.CREATED.statusCode)
        assertThat(emailArg.allValues.toSet()).isEqualTo(setOf(email))
        assertThat(skuArg.allValues.toSet()).isEqualTo(setOf(sku))
        assertThat(sourceIdArg.allValues.toSet()).isEqualTo(setOf(sourceId))
        assertThat(saveSourceArg.allValues.toSet()).isEqualTo(setOf(false))
    }

    companion object {

        val DAO: SubscriberDAO = mock(SubscriberDAO::class.java)
        val AUTHENTICATOR: OAuthAuthenticator = mock(OAuthAuthenticator::class.java)

        @JvmField
        @ClassRule
        val RULE = ResourceTestRule.builder()
                .setMapper(Jackson.newObjectMapper().registerModule(KotlinModule()))
                .addResource(AuthDynamicFeature(
                        OAuthCredentialAuthFilter.Builder<AccessTokenPrincipal>()
                                .setAuthenticator(AUTHENTICATOR)
                                .setPrefix("Bearer")
                                .buildAuthFilter()))
                .addResource(AuthValueFactoryProvider.Binder(AccessTokenPrincipal::class.java))
                .addResource(ProductsResource(DAO))
                .setTestContainerFactory(GrizzlyWebTestContainerFactory())
                .build()
    }
}
